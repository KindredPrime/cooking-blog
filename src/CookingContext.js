import React from 'react';

const CookingContext = React.createContext({
  user: null
});

export default CookingContext;