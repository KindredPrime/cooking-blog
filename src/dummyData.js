const dummyPosts = [
  {
    id: 1,
    lastEdited: 'Dec 17 07:37:16 2020',
    title: 'Lessons I Learned from Cooking',
    author: 'FakeUser2',
    content: `Have patience, it can be fun to experiment and explore, having a hard time learning something may only be temporary, bringing an idea of yours to life by creating it from scratch without relying on any instructions can be fun and rewarding, the meals you cook don't have to be fancy or difficult to taste good.`
  },
  {
    id: 2,
    lastEdited: 'Dec 28 18:24:43 2020',
    title: 'How to Save Money When Cooking',
    author: 'User1',
    content: `Buy in bulk, cook in bulk and freeze the leftovers, but store brand products.`
  },
  {
    id: 3,
    lastEdited: 'Dec 1 12:54:31 2020',
    title: '5 Easy Cooking Tips',
    author: 'User3',
    content: `1. When experimenting with seasonings in a recipe, mix them dry in a cup and smell it. If it smells good, then it'll probably taste good.
    2. Add some water when microwaving leftovers
    3. Start small with seasonings
    4. Cook
    5. Blah`
  },
  {
    id: 4,
    lastEdited: 'Jan 5 08:28:28 2021',
    title: 'Cooking Basics for Beginners',
    author: 'User1',
    content: `Go easy on seasonings and salt, if the inside of cooked chicken is pink or its juices aren't clear then it needs to be cooked longer, cooked ground beef needs to be brown in order to be safe to eat`
  },
  {
    id: 5,
    lastEdited: 'Dec 15 11:56:48 2020',
    title: 'Winter Seasoning Guide',
    author: 'User 2',
    content: `Rosemary, sage, thyme, cinnamon, nutmeg`
  },
  {
    id: 6,
    lastEdited: 'Jan 5 14:22:39 2021',
    title: 'Title6',
    author: 'User3',
    content: 'Post 6 content'
  },
  {
    id: 7,
    lastEdited: 'Jan 2 16:22:39 2021',
    title: 'Title7',
    author: 'User3',
    content: 'Post 7 content'
  },
  {
    id: 8,
    lastEdited: 'Jan 6 12:22:39 2021',
    title: 'Title8',
    author: 'User3',
    content: 'Post 8 content'
  },
  {
    id: 9,
    lastEdited: 'Dec 15 11:22:39 2021',
    title: 'Title9',
    author: 'User3',
    content: 'Post 9 content'
  },
  {
    id: 10,
    lastEdited: 'Dec 15 19:22:39 2021',
    title: 'Title10',
    author: 'User3',
    content: 'Post 10 content'
  },
  {
    id: 11,
    lastEdited: 'Jan 4 18:22:39 2021',
    title: 'Title11',
    author: 'User3',
    content: 'Post 11 content'
  }
];

const dummyComments = [
  {
    id: 1,
    lastEdited: new Date('Dec 19 07:37:16 2020'),
    content: 'These are all great lessons!',
    creator: 'User1',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 2,
    lastEdited: new Date('Dec 17 08:37:52 2020'),
    content: `I can't wait to try these seasoning combinations in my next meal!`,
    creator: 'User1',
    postTitle: 'Winter Seasoning Guide'
  },
  {
    id: 3,
    lastEdited: new Date('Jan 5 08:28:28 2021'),
    content: `I wish I knew these when I started learning to cook!`,
    creator: 'User4',
    postTitle: 'Cooking Basics for Beginners'
  },
  {
    id: 4,
    lastEdited: new Date('Dec 31 09:24:43 2020'),
    content: `Great advice!  This'll help me save so much money feeding my three teenagers!`,
    creator: 'User4',
    postTitle: `How to Save Money When Cooking`
  },
  {
    id: 5,
    lastEdited: new Date('Jan 1 10:31:39 2021'),
    content: `What an insightful post`,
    creator: 'FakeUser2',
    postTitle: 'Title7'
  },
  {
    id: 6,
    lastEdited: new Date('Dec 19 08:37:16 2020'),
    content: 'I know, right?',
    creator: 'User3',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 7,
    lastEdited: new Date('Dec 20 10:43:16 2020'),
    content: `It's amazing what you can learn from cooking`,
    creator: 'User4',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 8,
    lastEdited: new Date('Dec 17 20:12:16 2020'),
    content: 'Huh, I never thought about it this way',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 9,
    lastEdited: new Date('Dec 12 07:07:07 2020'),
    content: 'Comment9',
    creator: 'User6',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 10,
    lastEdited: new Date('Dec 19 15:22:16 2020'),
    content: 'Comment10',
    creator: 'User7',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 11,
    lastEdited: new Date('Dec 15 17:18:19 2020'),
    content: 'Comment11',
    creator: 'User8',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 12,
    lastEdited: new Date('Jan 1 12:13:14 2021'),
    content: 'Comment12',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 13,
    lastEdited: new Date('Jan 2 13:14:15 2021'),
    content: 'Comment13',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 14,
    lastEdited: new Date('Jan 5 06:07:08 2021'),
    content: 'Comment14',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 15,
    lastEdited: new Date('Jan 3 20:21:22 2021'),
    content: 'Comment15',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  }
];

const dummyUsers = [
  {
    id: 1,
    username: 'User1',
    password: 'password1',
    email: 'user1@gmail.com'
  },
  {
    id: 2,
    username: 'FakeUser2',
    password: 'password2',
    email: 'fakeuser2@gmail.com'
  },
  {
    id: 3,
    username: 'User3',
    password: 'password3',
    email: 'user3@gmail.com'
  },
  {
    id: 4,
    username: 'User4',
    password: 'password4',
    email: 'user4@gmail.com'
  },
  {
    id: 5,
    username: 'User5',
    password: 'password5',
    email: 'user5@gmail.com'
  },
  {
    id: 6,
    username: 'User6',
    password: 'password6',
    email: 'user6@gmail.com'
  },
  {
    id: 7,
    username: 'User7',
    password: 'password7',
    email: 'user7@gmail.com'
  },
  {
    id: 8,
    username: 'User8',
    password: 'password8',
    email: 'user8@gmail.com'
  }
];

export {
  dummyPosts,
  dummyComments,
  dummyUsers
};