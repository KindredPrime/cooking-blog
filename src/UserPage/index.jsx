import { Component } from 'react';
import PropTypes from 'prop-types';
import CookingContext from '../CookingContext';
import API from '../apiCalls';
import APIError from '../APIError';
import PostsList from '../PostsList/index';
import UserCommentsList from '../UserCommentsList/index';

class UserPage extends Component {
  static contextType = CookingContext;

  state = {
    username: null,
    email: null,
    blogPosts: null,
    comments: null,
    error: null
  };

  componentDidMount() {
    const { id } = this.props.match.params;
    const { user } = this.context;

    let userId;
    API.getUserById(id)
      .then((pageUser) => {
        userId = pageUser.id;

        if (user && pageUser.id === user.id) {
          this.setState({
            email: pageUser.email
          });
        }

        this.setState({
          username: pageUser.username,
          error: null
        });
      })
      .then(() => API.getCommentsByCreator(userId))
      .then((comments) => {
        this.setState({ comments })
      })
      .then(() => API.getBlogPostsByAuthor(userId))
      .then((blogPosts) => {
        this.setState({ blogPosts })
      })
      .catch((error) => {
        this.setState({
          error
        });
      });
  }

  componentWillUnmount() {
    API.abortTasks();
  }

  render() {
    const { username, email, blogPosts, comments, error } = this.state;

    return (
      <main className="UserPage">
        <header>
          <h1>{username}</h1>
        </header>

        {email && <p>Email: {email}</p>}

        {error && <APIError message={error.message} />}

        {blogPosts && <PostsList initialBlogPosts={blogPosts} />}

        {comments && <UserCommentsList initialComments={comments} />}
      </main>
    );
  }
}

UserPage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired
  }).isRequired
};

export default UserPage;