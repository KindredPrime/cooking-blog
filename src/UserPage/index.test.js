import ReactDOM from 'react-dom';
import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import API from '../apiCalls';
import UserPage from './index';
import { clientBlogPosts, clientComments, apiUsers } from '../dummyData';
import CookingContext from '../CookingContext';

describe('UserPage Component', () => {
  const origAPI = {
    getUserById: API.getUserById,
    getCommentsByCreator: API.getCommentsByCreator,
    getBlogPostsByAuthor: API.getBlogPostsByAuthor
  };

  // Mock API calls
  beforeAll(() => {
    API.getUserById = (id) => {
      return Promise.resolve(apiUsers.find((user) => user.id === parseInt(id)));
    };

    API.getCommentsByCreator = (id) => {
      return Promise.resolve(clientComments.filter((comment) => comment.creatorId === id));
    };

    API.getBlogPostsByAuthor = (id) => {
      return Promise.resolve(clientBlogPosts.filter((post) => post.authorId === id));
    };
  });

  afterAll(() => {
    API.getUserById = origAPI.getUserById;
    API.getCommentsByCreator = origAPI.getCommentsByCreator;
    API.getBlogPostsByAuthor = origAPI.getBlogPostsByAuthor;
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <UserPage match={{ params: { id: '1' }}} />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', async () => {
    const id = 3;
    const testUser = apiUsers[id-1];
    const contextValue = {
      user: testUser
    };
    render(
      <BrowserRouter>
        <CookingContext.Provider value={contextValue}>
          <UserPage match={{ params: { id: id.toString() }}} />
        </CookingContext.Provider>
      </BrowserRouter>
    );

    await waitFor(() => expect(screen.getByText(testUser.username)).toBeInTheDocument());

    expect(document.body).toMatchSnapshot();
  });

  it('renders an error as expected', async () => {
    const error = 'Test Error';
    const tempOrigAPI = {
      getUserById: API.getUserById
    };
    API.getUserById = () => new Promise((resolve, reject) => {
      throw new Error(error);
    });

    render(
      <BrowserRouter>
        <UserPage match={{ params: { id: '1' } }} />
      </BrowserRouter>
    );

    await waitFor(() => expect(screen.getByText(error)).toBeInTheDocument());
    expect(document.body).toMatchSnapshot();

    // Reset the API
    API.getUserById = tempOrigAPI.getUserById;
  });
});