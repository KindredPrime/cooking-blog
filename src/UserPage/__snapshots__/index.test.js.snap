// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPage Component renders an error as expected 1`] = `
<body>
  <div>
    <main
      class="UserPage"
    >
      <header>
        <h1 />
      </header>
      <p
        class="APIError"
      >
        Test Error
      </p>
    </main>
  </div>
</body>
`;

exports[`UserPage Component renders the UI as expected 1`] = `
<body>
  <div>
    <main
      class="UserPage"
    >
      <header>
        <h1>
          User3
        </h1>
      </header>
      <p>
        Email: 
        user3@gmail.com
      </p>
      <section
        class="PostsList"
      >
        <header>
          <h2>
            Posts
          </h2>
        </header>
        <ul>
          <li
            class="PostsList__blog-post"
          >
            <a
              class="PostsList__title"
              href="/blog-posts/8"
            >
              Title8
            </a>
            <p
              class="timestamp"
            >
              Last edited: 
              Jan 06 2021 12:22:39
            </p>
          </li>
          <li
            class="PostsList__blog-post"
          >
            <a
              class="PostsList__title"
              href="/blog-posts/6"
            >
              Title6
            </a>
            <p
              class="timestamp"
            >
              Last edited: 
              Jan 05 2021 14:22:39
            </p>
          </li>
          <li
            class="PostsList__blog-post"
          >
            <a
              class="PostsList__title"
              href="/blog-posts/11"
            >
              Title11
            </a>
            <p
              class="timestamp"
            >
              Last edited: 
              Jan 04 2021 18:22:39
            </p>
          </li>
          <li
            class="PostsList__blog-post"
          >
            <a
              class="PostsList__title"
              href="/blog-posts/7"
            >
              Title7
            </a>
            <p
              class="timestamp"
            >
              Last edited: 
              Jan 02 2021 16:22:39
            </p>
          </li>
          <li
            class="PostsList__blog-post"
          >
            <a
              class="PostsList__title"
              href="/blog-posts/10"
            >
              Title10
            </a>
            <p
              class="timestamp"
            >
              Last edited: 
              Dec 15 2020 19:22:39
            </p>
          </li>
          <li
            class="PostsList__blog-post"
          >
            <a
              class="PostsList__title"
              href="/blog-posts/9"
            >
              Title9
            </a>
            <p
              class="timestamp"
            >
              Last edited: 
              Dec 15 2020 11:22:39
            </p>
          </li>
          <li
            class="PostsList__blog-post"
          >
            <a
              class="PostsList__title"
              href="/blog-posts/3"
            >
              5 Easy Cooking Tips
            </a>
            <p
              class="timestamp"
            >
              Last edited: 
              Dec 01 2020 12:54:31
            </p>
          </li>
        </ul>
      </section>
      <section
        class="UserCommentsList"
      >
        <header>
          <h2>
            Comments
          </h2>
        </header>
        <ul>
          <li
            class="Comment"
          >
            <a
              class="Comment__post-title"
              href="/blog-posts/1"
            >
              Lessons I Learned from Cooking
            </a>
            <p
              class="Comment__content"
            >
              I know, right?
            </p>
            <p
              class="timestamp"
            >
              Last edited: 
              Dec 19 2020 08:37:16
            </p>
          </li>
        </ul>
      </section>
    </main>
  </div>
</body>
`;
